src/com/google/enterprise/adaptor/testing/RecordingDocIdPusher.java

```diff
@@ -12,57 +12,93 @@
+import com.google.enterprise.adaptor.ExceptionHandler;
+import com.google.enterprise.adaptor.GroupPrincipal;
+import com.google.enterprise.adaptor.Principal;
 
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.TreeMap;
 
-class AccumulatingDocIdPusher extends AbstractDocIdPusher {
+/**
+ * A fake implementation of {@link DocIdPusher} that simply records
+ * the values it receives.
+ */
+public class RecordingDocIdPusher extends AbstractDocIdPusher {
   private List<DocId> ids = new ArrayList<DocId>();
-  private List<DocIdPusher.Record> records
-      = new ArrayList<DocIdPusher.Record>();
+  private List<Record> records = new ArrayList<Record>();
+  private Map<DocId, Acl> namedResources = new TreeMap<DocId, Acl>();
   private Map<GroupPrincipal, Collection<Principal>> groups
       = new TreeMap<GroupPrincipal, Collection<Principal>>();
 
   @Override
   public Record pushRecords(Iterable<Record> records, ExceptionHandler handler)
       throws InterruptedException {
     for (Record record : records) {
-      this.records.add(record);
       ids.add(record.getDocId());
+      this.records.add(record);
     }
     return null;
   }
 
-  public List<DocId> getDocIds() {
-    return Collections.unmodifiableList(ids);
-  }
-
-  public List<DocIdPusher.Record> getRecords() {
-    return Collections.unmodifiableList(records);
-  }
-
-  public void reset() {
-    ids.clear();
-    records.clear();
-  }
-
   @Override
   public DocId pushNamedResources(Map<DocId, Acl> resources,
       ExceptionHandler handler) throws InterruptedException {
-    throw new UnsupportedOperationException();
+    namedResources.putAll(resources);
+    return null;
   }
 
   @Override
   public GroupPrincipal pushGroupDefinitions(
       Map<GroupPrincipal, ? extends Collection<Principal>> defs,
       boolean caseSensitive, ExceptionHandler handler)
       throws InterruptedException {
-    groups.putAll(defs); 
+    for (GroupPrincipal key : defs.keySet()) {
+      Collection<Principal> members = defs.get(key);
+      if (members instanceof List) {
+        groups.put(key, unmodifiableList(new ArrayList<Principal>(members)));
+      } else if (members instanceof Set) {
+        groups.put(key, unmodifiableSet(new LinkedHashSet<Principal>(members)));
+      } else {
+        groups.put(key, members);
+      }
+    }
     return null;
   }
+
+  public List<DocId> getDocIds() {
+    return Collections.unmodifiableList(ids);
+  }
+
+  public List<Record> getRecords() {
+    return Collections.unmodifiableList(records);
+  }
+
+  public Map<DocId, Acl> getNamedResources() {
+    return Collections.unmodifiableMap(namedResources);
+  }
+
+  public Map<GroupPrincipal, Collection<Principal>> getGroupDefinitions() {
+    return Collections.unmodifiableMap(groups);
+  }
+
+  public void reset() {
```

Comment:
I have two followups for this CL, one to add javadoc, and the other to propose some semantic changes from what we have in the library and connectors at the moment. I don't know why all of these had a reset method, but I'll look at that. I think there are tests that don't use reset or get a new DocIdPusher, and the assertions look funny, since they include previously collected and asserted results as well.